### PR TITLE
Adjust Copy-SqlJob to not copy jobs associated with Maintenance Plans

### LINKED ISSUE #126: Copy-SqlJob does not copy maintenance plans
If you are dealing with maintenance plans (not custom SSIS packages) for a given instance the cmdlet is only copying the job definition, not the actual maintenance plan.

Do we want cmdlet for copying maintenance plans or include T-SQL to copy msdb.dbo.sysssispackages (just maintenance plans) to destination. Other option would be to look at using SMO for SSIS to export the packages, but T-SQL likely going to be much less painful. Possibly this code: http://dba.stackexchange.com/a/112434/507

### PR DESCRIPTION
**Pull request by wsmelton**
Added `-IgnoreMaintenancePlan` to function so it will offer ability to ignore maintenance plans as we do not currently support moving those in the migration process.

Will resolve #126 

### General tests

- [ ] Copy maintenance plan from SQL 2000 to 2008 R2 and 2012, 2014 and 2016
- [ ] Copy maintenance plan from SQL 2005 to 2008 R2 and 2012, 2014 and 2016
- [x] Copy maintenance plan from 2008 R2 to 2012, 2014 and 2016
- [x] Copy maintenance plan from 2012 to 2014 and 2016
- [x] Copy maintenance plan from 2014 to 2016
- [x] `Copy-SqlJob` **without** the `-IncludeMaintenancePlan` switch, job associated with Maintenance Plan _is not_ copied over
- [x] `Copy-SqlJob` **with** the `-IncludeMaintenancePlan` switch, job associated with Maintenance Plan _is_ copied over, Maintenance Plan is copied over.

### Tests for removing Maintenance Plans

Manually remove Maintenance Plan from destination, ensure associated SQL Agent job is removed:
- [x] SQL 2008 R2
- [x] SQL 2012
- [x] SQL 2014
- [x] SQL 2016

If the SQL Agent job is created by editing the Maintenance Plan on the destination, it is removed successfully when the Maintenance Plan is removed. If the job was copied over using `Copy-SqlJob`, then it remains when the user deletes the Maintenance Plan.

### Has been tested on minimum requirements:
- [ ]  PowerShell 3
- [x]  Windows 7
- [ ]  SQL Server 2000 (does not currently work)

```PowerShell
Exception calling "ExecuteWithResults" with "1" argument(s): "An exception occurred while executing a Transact-SQL statement or batch."
At C:\Users\DanielAlexander\My Documents\WindowsPowerShell\Modules\dbatools\internal\DynamicParams.ps1:1994 char:2
```

- [ ] SQL Server 2005 (does not currently work)

```PowerShell
C:\Users\DanielAlexander\Documents\WindowsPowerShell\Modules\dbatools\functions\Copy-SqlMaintenancePlan.ps1
At C:\Users\DanielAlexander\Documents\WindowsPowerShell\Modules\dbatools\functions\Copy-SqlMaintenancePlan.ps1:162 char:6
+ ...             $MaintMetaData = $sourceserver.ConnectionContext.ExecuteW ...
+                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
System.Management.Automation.MethodInvocationException: Exception calling "ExecuteWithResults" with "1" argument(s): "An exception occurred while executing a Transact-SQL statement or batch." ---> Microsoft.SqlServer.Management.Common.ExecutionFailureException: An exception occurred while executing a Transact-SQL statement or batch. ---> System.Data.SqlClient.SqlException: Invalid object name 'msdb.dbo.sysssispackages'.
   at Microsoft.SqlServer.Management.Common.ConnectionManager.ExecuteTSql(ExecuteTSqlAction action, Object execObject, DataSet fillDataSet, Boolean catchException)
   at Microsoft.SqlServer.Management.Common.ServerConnection.ExecuteWithResults(String sqlCommand)
   --- End of inner exception stack trace ---
   at Microsoft.SqlServer.Management.Common.ServerConnection.ExecuteWithResults(String sqlCommand)
   at CallSite.Target(Closure , CallSite , Object , Object )
   --- End of inner exception stack trace ---
   at System.Management.Automation.ExceptionHandlingOps.CheckActionPreference(FunctionContext funcContext, Exception exception)
   at System.Management.Automation.Interpreter.ActionCallInstruction`2.Run(InterpretedFrame frame)
   at System.Management.Automation.Interpreter.EnterTryCatchFinallyInstruction.Run(InterpretedFrame frame)
   at System.Management.Automation.Interpreter.EnterTryCatchFinallyInstruction.Run(InterpretedFrame frame)
```

### Has been tested on maximum requirements:
- [x] PowerShell 5
- [ ]  Windows 10
- [ ]  SQL Server vNext

## QA checks
- [x] Working/useful help content, including link to command on dbatools web site
- [x] All examples work as advertised
- [x] Does not contain template content
- [x] Does not contain excessive/unnecessary amounts of comments
- [x] Works remotely
- [ ] Works locally
- [ ] Works on lower versions or throws error specifying version not supported
- [x] Works with named instances
- [x] Works with clustered instances
- [ ] No un-handled errors which stop the command working with multiple servers

## QA comments

- Notation within the maintenance plan seems to be lost, so far, in tests from 2008 R2 to other versions, where is this information stored? Is this a big deal, what else might be lost that I haven't tested for?
- Does not currently work with SQL 2000 due to an error in `Get-ParamSqlMaintenancePlans`, the user is not presented with a warning
- Does not currently work with SQL 2005 due to an error in `Copy-SqlMaintenancePlan`, the user is not presented with a warning

### CODE REVIEW

**Comment by dansqldba** [comment]
May be able to squeeze this one into the September release. Nice work, looking forward to testing this one!

**Comment by wsmelton** [comment]
@dansqldba @ctrlbold 

I found after interesting issue in client environment once I had the maintenance plan copied over, after the job was copied over, that caused unrecoverable issue. I basically had to recreate the maintenance plans in that environment with new names. The maintenance plans that got copied over would not run because the local connection in the plan never updated correctly...best guess this was from the associated job being there already.

I'd think it be best if by default the `Copy-SqlJob` does not copy any job associated with a maintenance plan, unless the user includes the flag to "IncludeMaintenancePlan".

**Comment by schwartzmx** [comment]
Hi @wsmelton,  I too had that problem when manually exporting/importing maintenance plans to another server. 
>The maintenance plans that got copied over would not run because the local connection in the plan never updated correctly...

I used this to alter the local connection string, in place so I didn't have to recreate the plans
```sql
	USE [msdb]
	GO
	DECLARE @OldServerName NVARCHAR(20) = 'Source-SQL-Server'
			,@NewServerName NVARCHAR(20)= 'Dest-SQL-Server'

	;WITH Packages AS (
		SELECT  x.*,
				LocalServerConnectionString = cm.value('declare namespace DTS="www.microsoft.com/SqlServer/Dts";DTS:ObjectData[1]/DTS:ConnectionManager[1]/@DTS:ConnectionString', 'varchar(1000)')
		FROM (
			SELECT  id, name, packageXML = CAST(CAST(packagedata AS VARBINARY(MAX)) AS XML)
			FROM dbo.sysssispackages
			WHERE id IN (SELECT id FROM dbo.sysmaintplan_plans)
		) x
		CROSS APPLY packageXML.nodes('declare namespace DTS="www.microsoft.com/SqlServer/Dts";/DTS:Executable/DTS:ConnectionManagers/DTS:ConnectionManager[@DTS:ObjectName="Local server connection"]') p(cm)
	)
	UPDATE dbo.sysssispackages 
	SET packagedata = CAST(CAST(REPLACE(CAST(CAST(packagedata AS VARBINARY(MAX)) AS VARCHAR(MAX)), @OldServerName, @NewServerName) AS XML) AS VARBINARY(MAX))
	WHERE id IN (select id from Packages)
	GO

	-- Verify
	SELECT  x.*,
			LocalServerConnectionString = cm.value('declare namespace DTS="www.microsoft.com/SqlServer/Dts";DTS:ObjectData[1]/DTS:ConnectionManager[1]/@DTS:ConnectionString', 'varchar(1000)')
	FROM (
		SELECT  id, name, packageXML = CAST(CAST(packagedata AS VARBINARY(MAX)) AS XML)
		FROM dbo.sysssispackages
		WHERE id IN (SELECT id FROM dbo.sysmaintplan_plans)
	) x
	CROSS APPLY packageXML.nodes('declare namespace DTS="www.microsoft.com/SqlServer/Dts";/DTS:Executable/DTS:ConnectionManagers/DTS:ConnectionManager[@DTS:ObjectName="Local server connection"]') p(cm)
	GO
```
May be able to incorporate that in a future version with another flag.

Anyways, nice work!  Lgtm

**Comment by dansqldba** [comment]
@wsmelton would you like to incorporate the feedback from @schwartzmx before this cmdlet goes through QA?

**Comment by wsmelton** [comment]
I don't, the fix for that issue was to change how the command copied jobs over and that has been donein this PR already. It will ignore any maintenance plan associated job and only copy it when you want to copy the maintenance plan as well.

**Comment by ctrlbold** [comment]
Nice solution, @wsmelton. Is it a final or should we file an issue for Copy-SqlJob not to copy associated maint plan jobs by default?

**Comment by wsmelton** [comment]
@ctrlbold This [includes update to the `Copy-SqlJob`](https://github.com/sqlcollaborative/dbatools/pull/205/commits/6e2412aaa3874b7a54a6121d28016ccfd5bbc95e) and I went ahead and set it to by default not to copy maintenance plans. This modified a parameter to be `IncludeMaintenancePlan` instead of `Ignore` parameter. So by default it will not touch any of the maintenance plans. Guess I should have put that as a separate PR.

**Comment by ctrlbold** [comment]
Oh, awesome! Thank you, k.

**Comment by ctrlbold** [comment]
- [ ] Working/useful help content, including link to command on dbatools web site
- [ ] All examples work as advertised
- [ ] Does not contain template content
- [ ] Does not contain excessive/unnecessary amounts of comments
- [ ] Works remotely
- [ ] Works locally
- [ ] Works on lower versions or throws error specifying version not supported
- [ ] Works with named instances
- [ ] Works with clustered instances
- [ ] Handles offline/read only databases
- [ ] Supports multiple servers (at the command line or piped from Get-SqlRegisteredServerName)
- [ ] No un-handled errors which stop the command working with multiple servers

File: functions/Copy-SqlMaintenancePlan.ps1 [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+
+.PARAMETER DestinationSqlCredential
+Allows you to login to servers using SQL Logins as opposed to Windows Auth/Integrated/Trusted. To use:
+
+$dcred = Get-Credential, then pass this $dcred to the -DestinationSqlCredential parameter.
+
+Windows Authentication will be used if DestinationSqlCredential is not specified. SQL Server does not accept Windows credentials being passed as credentials.
+
+To connect as a different Windows user, run PowerShell as that user.
+
```
Comment (ctrlbold) [inline functions/Copy-SqlMaintenancePlan.ps1]: Add Help for Force

File: functions/Copy-SqlMaintenancePlan.ps1 [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+		$sourceserver = Connect-SqlServer -SqlServer $Source -SqlCredential $SourceSqlCredential
+		$destserver = Connect-SqlServer -SqlServer $Destination -SqlCredential $DestinationSqlCredential
+
+		$source = $sourceserver.DomainInstanceName
+		$destination = $destserver.DomainInstanceName
+	}
+	
+	PROCESS
+	{
+		$qryPlanNames = "SELECT plans.[name] AS plan_name, plans.id AS plan_id
```
Comment (ctrlbold) [inline functions/Copy-SqlMaintenancePlan.ps1]: Please use $sql when writing SQL. This makes SQL queries easier to find in all commands.

File: functions/Copy-SqlMaintenancePlan.ps1 [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+							 ),
+							 '$source','$destination')
+					   AS xml)
+					AS varbinary(max) ) AS packagedata,
+					packageformat, packagetype, vermajor, verminor, verbuild, vercomments, verid, isencrypted, readrolesid, writerolesid
+					FROM [msdb].[dbo].[sysssispackages]
+					WHERE name = '$planname'"
+
+					Write-Verbose "[SELECT] query - Maintenance metadata"
+					Write-Verbose $qryMaintMetaData
```
Comment (ctrlbold) [inline functions/Copy-SqlMaintenancePlan.ps1]: All $sql should be output in Write-Debug. We found that writing it in verbose was overwhelming.

File: functions/Copy-SqlMaintenancePlan.ps1 [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+					$qryMaintInsert = "INSERT INTO [msdb].[dbo].[sysssispackages] (
+						[name],[id],[description],[createdate],[folderid],[ownersid],[packagedata],[packageformat],[packagetype],[vermajor]
+						,[verminor],[verbuild],[vercomments],[verid],[isencrypted],[readrolesid],[writerolesid])
+						VALUES (@name,@id,@description,@createdate,
+								@folderid,@ownersid,
+								@packagedata,@packageformat,@packagetype,
+								@vermajor,@verminor,@verbuild,@vercomments,@verid,
+								@isencrypted,@readrolesid,@writerolesid)"
+					$cn = New-Object System.Data.SqlClient.SqlConnection
+					$cn.ConnectionString = $destserver.ConnectionContext.ConnectionString
```
Comment (ctrlbold) [inline functions/Copy-SqlMaintenancePlan.ps1]: Very nice :D

File: functions/DynamicParams.ps1 [hunk 5]
```diff
@@ -271,6 +271,55 @@ Function Get-ParamSqlLogins
 	return $newparams
 }
```
Comment (ctrlbold) [inline functions/DynamicParams.ps1]: Please remove from PR. I added i to DynamicParams with fixes; my bad, I didn't see it in this PR when I pulled your branch.

File: functions/DynamicParams.ps1 [hunk 6]
```diff
@@ -271,6 +271,55 @@ Function Get-ParamSqlLogins
...
+	# Reusable parameter setup
+	$newparams = New-Object System.Management.Automation.RuntimeDefinedParameterDictionary
+	$attributes = New-Object System.Management.Automation.ParameterAttribute
+
+	$attributes.ParameterSetName = "__AllParameterSets"
+	$attributes.Mandatory = $false
+	$attributes.Position = 4
+
+	# MaintenancePlan list parameter setup
+	if ($list) { $validationset = New-Object System.Management.Automation.ValidateSetAttribute -ArgumentList $list }
```
Comment (ctrlbold) [inline functions/DynamicParams.ps1]: If you're interested, it was this variable. You used $planList above but $list in this part. Everything else was good.

File: functions/Copy-SqlMaintenancePlan.ps1 [hunk 7]
```diff
@@ -0,0 +1,237 @@
...
+Copy-SqlMaintenancePlan -Source sqlserver2014a -Destination sqlcluster
+Copies all maintenance plans from sqlserver2014a to sqlcluster, using Windows credentials. If maintenance plans exist with the same name on sqlcluster, they will be skipped.
+
+.EXAMPLE   
+Copy-SqlMaintenancePlan -Source sqlserver2014a -Destination sqlcluster -Plan MyBackup -SqlCredential $cred -Force
+
+Copies a single maintenance plan, the MyBackup, from sqlserver2014a to sqlcluster, using SQL credentials for sqlserver2014a and Windows credentials for sqlcluster. If a plan with the same name exist on sqlcluster, it will be removed and recreated because -Force was used.
+
+.EXAMPLE   
+Copy-SqlMaintenancePlan -Source sqlserver2014a -Destination sqlcluster -WhatIf -Force
```
Comment (ctrlbold) [inline functions/Copy-SqlMaintenancePlan.ps1]: WhatIf should also show drop when drops occur, even when -Force is used. Solution provided below.

    PS C:\temp\dbatools> Copy-SqlMaintenancePlan -Source sqlserver2014a -Destination sqlcluster
    WARNING: [Plan: MaintenancePlan] exist at destination. Use -Force to drop and migrate.
    WARNING: [Plan: MaintenancePlan2] exist at destination. Use -Force to drop and migrate.
    Maintenance Plan migration finished
    PS C:\temp\dbatools> Copy-SqlMaintenancePlan -Source sqlserver2014a -Destination sqlcluster -Force
    Dropping Maintenance Plan MaintenancePlan on SQLCLUSTER
    Copying Maintenance Plan MaintenancePlan
    Dropping Maintenance Plan MaintenancePlan2 on SQLCLUSTER
    Copying Maintenance Plan MaintenancePlan2
    Maintenance Plan migration finished
    PS C:\temp\dbatools> Copy-SqlMaintenancePlan -Source sqlserver2014a -Destination sqlcluster -WhatIf -Force
    What if: Performing the operation "Creating Maintenance Plan: MaintenancePlan" on target "SQLCLUSTER".
    What if: Performing the operation "Creating Maintenance Plan: MaintenancePlan2" on target "SQLCLUSTER".
    What if: Performing the operation "Showing final message" on target "console".

Comment (ctrlbold) [inline functions/Copy-SqlMaintenancePlan.ps1]: This is weird.

    PS C:\temp\dbatools> Copy-SqlMaintenancePlan -Source sqlserver2014a -Destination sqlcluster -Exclude MaintenancePlan
    Copying Maintenance Plan MaintenancePlan2
    WARNING: See error log C:\Users\ctrlb.BASE\Documents\dbatools-exceptions.txt for more details.
    Maintenance Plan migration finished
    PS C:\temp\dbatools> C:\Users\ctrlb.BASE\Documents\dbatools-exceptions.txt
    PS C:\temp\dbatools> Copy-SqlMaintenancePlan -Source sqlserver2014a -Destination sqlcluster -Exclude MaintenancePlan
    Copying Maintenance Plan MaintenancePlan2
    WARNING: See error log C:\Users\ctrlb.BASE\Documents\dbatools-exceptions.txt for more details.
    Maintenance Plan migration finished
    PS C:\temp\dbatools> C:\Users\ctrlb.BASE\Documents\dbatools-exceptions.txt
    PS C:\temp\dbatools> Copy-SqlMaintenancePlan -Source sqlserver2014a -Destination sqlcluster
    Copying Maintenance Plan MaintenancePlan
    WARNING: See error log C:\Users\ctrlb.BASE\Documents\dbatools-exceptions.txt for more details.
    Copying Maintenance Plan MaintenancePlan2
    WARNING: See error log C:\Users\ctrlb.BASE\Documents\dbatools-exceptions.txt for more details.

Exception log says

System.Management.Automation.MethodInvocationException: Exception calling "ExecuteNonQuery" with "0" argument(s): "Violation of PRIMARY KEY constraint 'pk_sysssispackages'. Cannot insert duplicate key in object 'dbo.sysssispackages'. The duplicate key value is (08aa12d5-8f98-4dab-a4fc-980b150a5dc8, MaintenancePlan2).

But it worked at first?

File: functions/Copy-SqlMaintenancePlan.ps1 [hunk 8]
```diff
@@ -0,0 +1,237 @@
...
+			{ continue }
+
+			if ($destMaintPlanList.plan_name -contains $plans)
+			{
+				if ($force -eq $false)
+				{
+					Write-Warning "[Plan: $planname] exist at destination. Use -Force to drop and migrate."
+					continue
+				}
+				else
```
Comment (ctrlbold) [inline functions/Copy-SqlMaintenancePlan.ps1]: if this else is removed, then WhatIf and Force used together should work.

**Comment by wsmelton** [comment]
Removing the Copy-SqlJob modification from this PR to prevent merge conflict, and it should be a separate PR anyway.

File: functions/Copy-SqlMaintenancePlan.ps1 (hunk 7 above)
Comment (wsmelton) [inline functions/Copy-SqlMaintenancePlan.ps1]: @ctrlbold ok, think latest commit should fix all of this...one stupid variable change after looking at it for 3 days. 🤕

**Comment by wsmelton** [comment]
Went ahead and added the Copy-SqlJob adjustment back to this PR...seemed to big of a hassle to try and and create another branch just for that change.

In my testing though I do see plans and jobs copied, but in deleting the plan the associated job is not being dropped like it should. I'm not sure why it does that nor if we need to address before pushing this command out.

**Comment by wsmelton** [comment]
@dansqldba Discussed with Chrissy and right now it will be best if we just do not support Maintenance Plans at all. I tried to work with the API for SSIS but it does not seem to support just copying from one server to another, at least in a clean manner.

I've adjusted the PR to just have an adjustment to Copy-SqlJob so it ignores jobs associated with Maintenance Plans. So test should be simply include a job with a maintenance plan, and the command should ignore that job completely. It will output a warning that the job is associated with a plan and being skipped.

**Pull request merged**